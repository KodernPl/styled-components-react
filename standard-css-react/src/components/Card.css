@import '../styles/variables.css';

.card {
  width: 26rem;
  height: 47rem;
  perspective: 150rem;
  position: relative;

  @media (--desktop) {
    height: 44rem;

    &:hover &__side__front {
      transform: rotateY(-180deg);
    }

    &:hover &__side__back {
      transform: rotateY(0);
    }
  }

  &__side {
    height: 100%;
    width: 100%;
    transition: all 1s ease;
    position: absolute;
    top: 0;
    left: 0;
    backface-visibility: hidden;
    border-radius: .5rem;
    overflow: hidden;
    box-shadow: .5rem 1rem 4rem var(--color-alpha-black);
    background: var(--color-white);

    &__front {
      &__heading-wrapper {
        margin: -.1rem;
        width: calc(100% + .2rem);
        height: 13.2rem;
      }

      &__buttons-bottom {
        text-align: center;

        @media (--desktop) {
          display: none;
        }
      }
    }

    &__back {
      transform: rotateY(180deg);

      &--dark {
        background: var(--gradient-dark);
      }

      &--light {
        background: var(--gradient-light);
      }

      &__buttons-centered {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        padding: .1rem;
      }
    }
  }
}
