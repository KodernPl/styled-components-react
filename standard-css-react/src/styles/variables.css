:root {
  --color-white: white;
  --color-gradient1: #667db6;
  --color-gradient2: #0082c8;
  --color-gradient3: #0082c8;
  --color-gradient4: #667db6;
  --color-white: #fff;
  --color-black: #000;
  --color-dark-grey: #222;
  --color-darkest-grey: #111;
  --color-grey: grey;
  --color-palevioletred: palevioletred;
  --color-sc-pink: rgb(219, 112, 147);
  --color-sc-orange: rgb(218, 163, 87);
  --color-alpha-black: rgba(0, 0, 0, 0.2);
  --color-alpha-grey: rgba(20, 20, 20, 0.1);
  --gradient-light: linear-gradient(to top right, rgb(219, 112, 147), rgb(218, 163, 87));
  --gradient-dark: linear-gradient(to top right, #667db6, #0082c8, #0082c8);
}

@custom-media --desktop (min-width: 1024px);
