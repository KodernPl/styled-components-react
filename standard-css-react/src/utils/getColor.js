/* globals window */
export const getColor = variable => window
  .getComputedStyle(window.document.body)
  .getPropertyValue(variable);
