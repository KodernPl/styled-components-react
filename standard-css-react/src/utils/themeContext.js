import React from 'react';

const { Consumer, Provider } = React.createContext();

// eslint-disable-next-line react/display-name
const withContext = () => Component => props => (
  <Consumer>
    {theme => <Component {...props} theme={theme} />}
  </Consumer>
);

export const withTheme = withContext();
export const ThemeProvider = Provider;
