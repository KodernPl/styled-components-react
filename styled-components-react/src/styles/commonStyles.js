import styled from 'styled-components';

export const CenterElements = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
`;

export const CustomLink = styled.a`
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  text-decoration: none;
  font-size: 1.2rem;
`;
