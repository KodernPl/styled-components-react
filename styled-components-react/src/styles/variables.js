export const COLORS = {
  gradient1: '#667db6',
  gradient2: '#0082c8',
  gradient3: '#0082c8',
  gradient4: '#667db6',

  white: '#fff',
  black: '#000',
  darkGrey: '#222',
  darkestGrey: '#111',

  grey: 'grey',
  palevioletred: 'palevioletred',

  scPink: 'rgb(219, 112, 147)',
  scOrange: 'rgb(218, 163, 87)',

  alphaBlack: 'rgba(0, 0, 0, 0.2)',
  alphaGrey: 'rgba(20, 20, 20, 0.1)',
};

export const THEMES = {
  light: 'light',
  dark: 'dark',
};

export const GRADIENTS = {
  light: `linear-gradient(to top right, ${COLORS.scPink}, ${COLORS.scOrange})`,
  dark: `linear-gradient(to top right, ${COLORS.gradient1}, ${COLORS.gradient2}, ${COLORS.gradient4})`,
};

export const BREAKPOINTS = {
  desktop: '(min-width: 1024px)',
};
