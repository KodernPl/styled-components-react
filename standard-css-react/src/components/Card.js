import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { getColor } from '../utils/getColor';
import { withTheme } from '../utils/themeContext';
import {
  THEMES,
} from '../styles/variables';

import Heading from './Heading';
import LinkButton, { LINK_BUTTON_VARIANTS } from './LinkButton';
import Details from './Details';
import Logo from './Logo';

import './Card.css';

const ns = 'card';
const displayName = 'Card';

const propTypes = {
  theme: PropTypes.string.isRequired,
};

const Card = ({ theme }) => {
  const sideFrontClassNames = cx(
    `${ns}__side`,
    `${ns}__side__front`,
    {
      [`${ns}__side__front--light`]: theme === THEMES.light,
      [`${ns}__side__front--dark`]: theme === THEMES.dark,
    },
  );

  const sideBackClassNames = cx(
    `${ns}__side`,
    `${ns}__side__back`,
    {
      [`${ns}__side__back--light`]: theme === THEMES.light,
      [`${ns}__side__back--dark`]: theme === THEMES.dark,
    },
  );

  const colorGradient2 = getColor('--color-gradient2');
  const colorGradient4 = getColor('--color-gradient4');
  const colorScPink = getColor('--color-sc-pink');
  const colorScOrange = getColor('--color-sc-orange');

  const bottomColor = theme === THEMES.light ? colorScPink : colorGradient4;
  const topColor = theme === THEMES.light ? colorScOrange : colorGradient2;

  return (
    <div className={ns}>
      <div className={sideFrontClassNames}>
        <div className={`${ns}__side__front__heading-wrapper`}>
          <Heading bottomColor={bottomColor} topColor={topColor} />
        </div>
        <Logo alt="styled-components logo" src="images/sc-logo.png" />
        <Details>
          <span>Automatic critical CSS </span>
          <span>No class name bugs</span>
          <span>Easier deletion of CSS</span>
          <span>Simple dynamic styling</span>
          <span>Painless maintenance</span>
          <span>Automatic vendor prefixing</span>
        </Details>
        <div className={`${ns}__side__front__buttons-bottom`}>
          <LinkButton href="https://www.styled-components.com/docs/" target="_blank" variant={LINK_BUTTON_VARIANTS.transparentDark}>Documentation</LinkButton>
        </div>
      </div>
      <div className={sideBackClassNames}>
        <div className={`${ns}__side__back__buttons-centered`}>
          <LinkButton href="https://github.com/styled-components/styled-components" target="_blank">Github</LinkButton>
          <LinkButton href="https://www.styled-components.com/docs/" target="_blank" variant={LINK_BUTTON_VARIANTS.transparent}>Documentation</LinkButton>
        </div>
      </div>
    </div>
  );
};

Card.displayName = displayName;
Card.propTypes = propTypes;

export default withTheme(Card);
