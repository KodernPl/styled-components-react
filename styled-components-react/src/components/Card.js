import React from 'react';
import PropTypes from 'prop-types';
import styled, { css, withTheme } from 'styled-components';
import styledTheming from 'styled-theming';

import Heading from './Heading';
import LinkButton, { LINK_BUTTON_VARIANTS } from './LinkButton';
import Details from './Details';
import Logo from './Logo';

import {
  COLORS, GRADIENTS, THEMES, BREAKPOINTS,
} from '../styles/variables';
import { CenterElements } from '../styles/commonStyles';

const gradient = styledTheming('mode', {
  [THEMES.light]: GRADIENTS.light,
  [THEMES.dark]: GRADIENTS.dark,
});

const Side = css`
  height: 100%;
  width: 100%;
  transition: all 1s ease;
  position: absolute;
  top: 0;
  left: 0;
  backface-visibility: hidden;
  border-radius: .5rem;
  overflow: hidden;
  box-shadow: .5rem 1rem 4rem ${COLORS.alphaBlack};
  background: ${COLORS.white};
`;

const SideFront = styled.div`
  ${Side};
`;

const SideBack = styled.div`
  ${Side}
  transform: rotateY(180deg);
  background: ${gradient};
`;

const CardWrapper = styled.div`
  width: 26rem;
  height: 47rem;
  perspective: 150rem;
  position: relative;

  @media ${BREAKPOINTS.desktop} {
    height: 44rem;

    &:hover ${SideFront} {
      transform: rotateY(-180deg);
    }

    &:hover ${SideBack} {
      transform: rotateY(0);
    }
  }
`;

const HeadingWrapper = styled.div`
  margin: -.1rem;
  width: calc(100% + .2rem);
  height: 13.2rem;
`;

const ButtonsCentered = styled(CenterElements)`
  padding: .1rem;
`;

const ButtonsBottom = styled.div`
  text-align: center;

  @media ${BREAKPOINTS.desktop} {
    display: none;
  }
`;

const displayName = 'Card';

const propTypes = {
  theme: PropTypes.shape({
    mode: PropTypes.string,
  }).isRequired,
};

const Card = ({ theme }) => {
  const bottomColor = theme.mode === THEMES.light ? COLORS.scPink : COLORS.gradient4;
  const topColor = theme.mode === THEMES.light ? COLORS.scOrange : COLORS.gradient2;

  return (
    <CardWrapper>
      <SideFront>
        <HeadingWrapper>
          <Heading bottomColor={bottomColor} topColor={topColor} />
        </HeadingWrapper>
        <Logo alt="styled-components logo" src="/images/sc-logo.png" />
        <Details>
          <span>Automatic critical CSS </span>
          <span>No class name bugs</span>
          <span>Easier deletion of CSS</span>
          <span>Simple dynamic styling</span>
          <span>Painless maintenance</span>
          <span>Automatic vendor prefixing</span>
        </Details>
        <ButtonsBottom>
          <LinkButton href="https://www.styled-components.com/docs/" target="_blank" variant={LINK_BUTTON_VARIANTS.transparentDark}>Documentation</LinkButton>
        </ButtonsBottom>
      </SideFront>
      <SideBack>
        <ButtonsCentered>
          <LinkButton href="https://github.com/styled-components/styled-components" target="_blank">Github</LinkButton>
          <LinkButton href="https://www.styled-components.com/docs/" target="_blank" variant={LINK_BUTTON_VARIANTS.transparent}>Documentation</LinkButton>
        </ButtonsCentered>
      </SideBack>
    </CardWrapper>
  );
};

Card.displayName = displayName;
Card.propTypes = propTypes;

export default withTheme(Card);
