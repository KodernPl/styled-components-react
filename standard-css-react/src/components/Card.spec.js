import React from 'react';

import { mount } from 'enzyme';

import { ThemeProvider } from '../utils/themeContext';
import { THEMES } from '../styles/variables';

import Card from './Card';

const mockColors = {
  '--color-gradient2': '#667db6',
  '--color-gradient4': '#0082c8',
  '--color-sc-pink': 'rgb(219, 112, 147)',
  '--color-sc-orange': 'rgb(218, 163, 87)',
};

jest.mock('../utils/getColor', () => ({
  getColor(variable) {
    return mockColors[variable];
  },
}));

const renderComponent = ({
  theme = THEMES.light,
}) => mount(
  <ThemeProvider value={theme}>
    <Card />
  </ThemeProvider>,
);

test('renders correct HTML', () => {
  expect(renderComponent({})).toMatchSnapshot();
});

[
  { theme: THEMES.light, bottomColor: mockColors['--color-sc-pink'], topColor: mockColors['--color-sc-orange'] },
  { theme: THEMES.dark, bottomColor: mockColors['--color-gradient4'], topColor: mockColors['--color-gradient2'] },
].forEach(({ theme, bottomColor, topColor }) => {
  test(`should Heading have correct props for ${theme} theme`, () => {
    const headingProps = renderComponent({ theme }).find('Heading').props();

    expect(headingProps.bottomColor).toEqual(bottomColor);
    expect(headingProps.topColor).toEqual(topColor);
  });
});
