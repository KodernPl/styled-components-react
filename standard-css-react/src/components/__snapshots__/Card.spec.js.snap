// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct HTML 1`] = `
<Component>
  <Card
    theme="light"
  >
    <div
      className="card"
    >
      <div
        className="card__side card__side__front card__side__front--light"
      >
        <div
          className="card__side__front__heading-wrapper"
        >
          <Heading
            bottomColor="rgb(219, 112, 147)"
            topColor="rgb(218, 163, 87)"
          >
            <svg
              height="100%"
              preserveAspectRatio="none"
              viewBox="0 0 100 100"
              width="100%"
            >
              <defs>
                <linearGradient
                  id="heading-linear-gradient"
                  x1="-1%"
                  x2="101%"
                  y1="100%"
                  y2="0%"
                >
                  <stop
                    offset="0%"
                    style={
                      Object {
                        "stopColor": "rgb(219, 112, 147)",
                      }
                    }
                  />
                  <stop
                    offset="100%"
                    style={
                      Object {
                        "stopColor": "rgb(218, 163, 87)",
                      }
                    }
                  />
                </linearGradient>
              </defs>
              <g>
                <path
                  d="M0 0 L0 100 L100 80 L100 0 Z"
                  fill="url(#heading-linear-gradient)"
                />
              </g>
            </svg>
          </Heading>
        </div>
        <Component
          alt="styled-components logo"
          src="images/sc-logo.png"
        >
          <Logo
            alt="styled-components logo"
            src="images/sc-logo.png"
            theme="light"
          >
            <img
              alt="styled-components logo"
              className="logo logo--light"
              src="images/sc-logo.png"
              theme="light"
            />
          </Logo>
        </Component>
        <Details>
          <ul
            className="details"
          >
            <li
              className="details__item"
              key="Automatic critical CSS "
            >
              <span>
                Automatic critical CSS 
              </span>
            </li>
            <li
              className="details__item"
              key="No class name bugs"
            >
              <span>
                No class name bugs
              </span>
            </li>
            <li
              className="details__item"
              key="Easier deletion of CSS"
            >
              <span>
                Easier deletion of CSS
              </span>
            </li>
            <li
              className="details__item"
              key="Simple dynamic styling"
            >
              <span>
                Simple dynamic styling
              </span>
            </li>
            <li
              className="details__item"
              key="Painless maintenance"
            >
              <span>
                Painless maintenance
              </span>
            </li>
            <li
              className="details__item"
              key="Automatic vendor prefixing"
            >
              <span>
                Automatic vendor prefixing
              </span>
            </li>
          </ul>
        </Details>
        <div
          className="card__side__front__buttons-bottom"
        >
          <Component
            href="https://www.styled-components.com/docs/"
            target="_blank"
            variant="transparentDark"
          >
            <LinkButton
              href="https://www.styled-components.com/docs/"
              target="_blank"
              theme="light"
              variant="transparentDark"
            >
              <a
                className="link-button link-button--transparentDark--light"
                href="https://www.styled-components.com/docs/"
                target="_blank"
                theme="light"
                variant="transparentDark"
              >
                Documentation
              </a>
            </LinkButton>
          </Component>
        </div>
      </div>
      <div
        className="card__side card__side__back card__side__back--light"
      >
        <div
          className="card__side__back__buttons-centered"
        >
          <Component
            href="https://github.com/styled-components/styled-components"
            target="_blank"
          >
            <LinkButton
              href="https://github.com/styled-components/styled-components"
              target="_blank"
              theme="light"
              variant="default"
            >
              <a
                className="link-button link-button--default--light"
                href="https://github.com/styled-components/styled-components"
                target="_blank"
                theme="light"
                variant="default"
              >
                Github
              </a>
            </LinkButton>
          </Component>
          <Component
            href="https://www.styled-components.com/docs/"
            target="_blank"
            variant="transparent"
          >
            <LinkButton
              href="https://www.styled-components.com/docs/"
              target="_blank"
              theme="light"
              variant="transparent"
            >
              <a
                className="link-button link-button--transparent--light"
                href="https://www.styled-components.com/docs/"
                target="_blank"
                theme="light"
                variant="transparent"
              >
                Documentation
              </a>
            </LinkButton>
          </Component>
        </div>
      </div>
    </div>
  </Card>
</Component>
`;
